File: src/Session.cpp
#include "Session.h"

JNI_METHOD_BEGIN(jlong, create)

    session *s = new session();

    return (jlong) s;

JNI_METHOD_END

JNI_METHOD_BEGIN(void, release, jlong handle)

    session *s = (session *) handle;

    delete s;

JNI_METHOD_END

JNI_METHOD_BEGIN(void, startUPnP, jlong handle)

    session *s = (session *) handle;

    s->start_upnp();

JNI_METHOD_END

JNI_METHOD_BEGIN(void, startNATPMP, jlong handle)

    session *s = (session *) handle;

    s->start_natpmp();

JNI_METHOD_END

JNI_METHOD_BEGIN(void, startLSD, jlong handle)

    session *s = (session *) handle;

    s->start_lsd();

JNI_METHOD_END

JNI_METHOD_BEGIN(void, startDHT, jlong handle)

    session *s = (session *) handle;

    s->start_dht();

JNI_METHOD_END

JNI_METHOD_BEGIN(jobjectArray, waitForAlerts, jlong handle, jint millis)

    session *s = (session *) handle;

    deque<alert *> alerts;

    if (s->wait_for_alert(milliseconds(millis)) != 0) {
        s->pop_alerts(&alerts);
    }

    for (alert *a: alerts) {
        cout << (*a).what() << "\n";
    }

JNI_METHOD_END

File: java/src/com/frostwire/libtorrent/Session.java
package com.frostwire.libtorrent;

import com.frostwire.libtorrent.alerts.Alert;

import java.util.Arrays;
import java.util.List;

public class Session {

    private final long h;

    public Session() {
        this.h = create();
    }

    public void startNetworking() {
        startUPnP(h);
        startNATPMP(h);
        startLSD(h);
        startDHT(h);
    }

    public void stopNetworking() {
        stopUPnP(h);
        stopNATPMP(h);
        stopLSD(h);
        stopDHT(h);
    }

    public List<Alert> waitForAlerts(int millis) {
        return Arrays.asList(waitForAlert(h, millis));
    }

    @Override
    protected void finalize() throws Throwable {
        if (this.h != 0) {
            release(this.h);
        }
        super.finalize();
    }

    private native long create();

    private native void release(long h);

    private native void startUPnP(long h);

    private native void startNATPMP(long h);

    private native void startLSD(long h);

    private native void startDHT(long h);

    private native void stopUPnP(long h);

    private native void stopNATPMP(long h);

    private native void stopLSD(long h);

    private native void stopDHT(long h);

    private native Alert[] waitForAlert(long h, int millis);
}
